Examples/CaseStudies/SwiftUICaseStudiesTests/01-GettingStarted-SharedStateTests.swift

```diff
@@ -0,0 +1,103 @@
+import Combine
+import ComposableArchitecture
+import XCTest
+
+@testable import SwiftUICaseStudies
+
+class SharedStateTests: XCTestCase {
+  func testTabRestoredOnReset() {
+    let store = TestStore(
+      initialState: SharedState(),
+      reducer: sharedStateReducer,
+      environment: ()
+    )
+    
+    store.assert(
+      .send(.selectTab(.profile)) {
+        $0.currentTab = .profile
+        $0.profile = .init(currentTab: .profile, count: 0, maxCount: 0, minCount: 0, numberOfCounts: 0)
+      },
+      .send(.profile(.resetCounterButtonTapped)) {
+        $0.currentTab = .counter
+        $0.profile = .init(currentTab: .counter, count: 0, maxCount: 0, minCount: 0, numberOfCounts: 0)
+      })
+  }
+
+  func testTabSelection() {
+    let store = TestStore(
+      initialState: SharedState(),
+      reducer: sharedStateReducer,
+      environment: ()
+    )
+
+    store.assert(
+      .send(.selectTab(.profile)) {
+        $0.currentTab = .profile
+        $0.profile = .init(currentTab: .profile, count: 0, maxCount: 0, minCount: 0, numberOfCounts: 0)
+      },
+      .send(.selectTab(.counter)) {
+        $0.currentTab = .counter
+        $0.profile = .init(currentTab: .counter, count: 0, maxCount: 0, minCount: 0, numberOfCounts: 0)
+      })
+  }
+
+  func testSharedCounts() {
+    let store = TestStore(
+      initialState: SharedState(),
+      reducer: sharedStateReducer,
+      environment: ()
+    )
+
+    store.assert(
+      .send(.counter(.incrementButtonTapped)) {
+        $0.counter = .init(alert: nil, count: 1, maxCount: 1, minCount: 0, numberOfCounts: 1)
```

Comment:
Is there a way to send an action to the store without having to assert the changes it makes? I was originally sending a few actions, and only checking the state on the last action, but found that the tests failed because each `send` step said the state had unexpectedly changed. As a result I opted to check the state for each action, which is arguably more correct, but it would still be nice to do such things.

Oddly I was able to bypass the test failure by only checking the state of the counter or the profile, I didn't have to do both - I'm not quite sure why that is (though I ended up checking both anyways since that seems more correct, and makes it more obvious that the two are mirroring one another).

By this I mean:
```
.send(.counter(.incrementButtonTapped)) //This by itself caused a test failure due to unexpected state change

.send(.counter(.incrementButtonTapped)) {
$0.counter = .init(alert: nil, count: 1, maxCount: 1, minCount: 0, numberOfCounts: 1)
} //This didn't have a test failure, even though the profile also changed but wasn't explicitly checked

.send(.counter(.incrementButtonTapped)) {
$0.profile = .init(currentTab: .counter, count: 1, maxCount: 1, minCount: 0, numberOfCounts: 1)
} //This also didn't have a test failure, even though I didn't explicitly check the counter state, which also changed.
```

I'm confused why I can't just omit the closure altogether, yet I can just omit one of the state checks.

Reply:
The reason you don't have to do both is because `profile` isn't real state held in the struct, it's just a computed property. Any mutations to it are immediately reflected in the `counter` state, and any mutations to `counter` are immediately reflected in `profile`.

It does make the assertion a little confusing, but at least the assertion is still being exhaustive in what it is checking.

> Is there a way to send an action to the store without having to assert the changes it makes? I was originally sending a few actions, and only checking the state on the last action, but found that the tests failed because each send step said the state had unexpectedly changed. As a result I opted to check the state for each action, which is arguably more correct, but it would still be nice to do such things.

We don't allow this because we want the `assert` helper to be as exhaustive as possible, and for you to really prove that you know how state is mutated each step of the way. Further, if a step ever executed an effect that was fed back into the system you would need to assert against that with a `.receive` step, and we wouldn't want you to be able to skip that either.

Reply:
Thank you for that clarification. I think this shows a bit of a flaw with doing shared state - it introduces a bit of "magic" that is hard to reason about. It's good that the architecture is able to support this, but I think I'd try hard to avoid using it in my own applications (or perhaps limit it to read only shared state, as opposed to bidirectional sharing, like the "reset profile" action does).

In regards to the assertions, I acknowledge that it is good to guide users towards more exhaustive tests, but that could also have the downside of users skipping some tests due to the hoops they would have to jump through to handle them.

Perhaps a future improvement could be an alternative less strict assert (maybe `softAssert`? `fuzzyAssert`? `unsafeAssert`?). That said, the testability of this architecture is a big reason why I became interested in it.